File: Java/src/com/senghuotlim/cci/LinkedList.java
package com.senghuotlim.cci;

public class LinkedList {
	
	public static void main(String[] args) {
		LinkedList head = new LinkedList();
		for (int i=1; i<=10; i++)
			head.add(i);
		System.out.println(head);
		
		head.reverse();
		System.out.println(head);
	}
	
	Node head;
	
	public void add(int data) {
		if (head == null)
			head = new Node(data, null);
		else {
			Node tmp = head;
			while (tmp.next != null)
				tmp = tmp.next;
			tmp.next = new Node(data, null);
		}
	}
	
	public String toString() {
		String res = "";
		Node tmp = head;
		while (tmp != null) {
			res += " " + tmp.data;
			tmp = tmp.next;
		}
		return res;
	}
	
	public void reverse() {
		Node reverse = reverse(head);
		head.next = null;
		head = reverse;
	}
	
	public Node reverse(Node head) {
		if (head == null || head.next == null)
			return head;
		
		Node next = head.next;
		Node reverse = reverse(head.next);
		next.next = head;
		
		return reverse;
	}
	
}

File: Python/linkedList.py
class LinkedList:
    def __init__(self):
        self.head = None

    def add(self, data):
        if self.head == None:
            self.head = Node(data)
        else:
            tmp = self.head
            while tmp.next != None:
                tmp = tmp.next
            tmp.next = Node(data)

    def toString(self):
        tmp = self.head
        res = ""
        while tmp != None:
            res += " " +  str(tmp.data)
            tmp = tmp.next
        return res

    def reverse(self):
        self.head = self.reverseHelper(self.head)

    def reverseHelper(self, head):
        if head == None or head.next == None:
            return head

        next = head.next
        reversed_list = self.reverseHelper(head.next)
        next.next = head
        head.next = None

        return reversed_list

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


list = LinkedList()
list.add(5)
list.add(6)
list.add(7)

print(list.toString())
list.reverse()
print(list.toString())